src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -1019,7 +1019,7 @@ internal static ConstantValue ExpressionOfTypeMatchesPatternType(
             ConstantValue result = Binder.GetIsOperatorConstantResult(expressionType, patternType, conversion.Kind, operandConstantValue, operandCouldBeNull);
 
             // Don't need to worry about checked user-defined operators
-            Debug.Assert(!conversion.IsUserDefined || result == ConstantValue.False || result == ConstantValue.Bad);
+            Debug.Assert((!conversion.IsUserDefined && !conversion.IsUnion) || result == ConstantValue.False || result == ConstantValue.Bad);
```

Comment:
These are checked together a lot, it might be helpful to declare property `IsUserDefinedOrUnion` to use in these places